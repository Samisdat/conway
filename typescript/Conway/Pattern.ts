import {CellMatrix} from '@Conway/Conway/CellMatrix';
import {Position} from '@Conway/Conway/Position';

export class Pattern {

    private name: String;

    private matrix: CellMatrix;

    constructor(name: String, matrix: CellMatrix) {

        this.name = name;

        this.matrix = matrix;

    }

    public getMatrix(): CellMatrix {

        return this.matrix;
    }

    public getWidth(): number {
        return this.matrix.width();
    }

    public getHeight(): number {
        return this.matrix.height();
    }

    public mirrorVertical(): void {

        const mirror = new CellMatrix();

        for (const position of this.matrix.all()) {

            mirror.add(
                new Position(
                    position.x,
                    -1 * position.y
                )
            );

        }

        this.matrix = mirror;

    }

    public mirrorHorizontal(): void {

        const mirror = new CellMatrix();

        for (const position of this.matrix.all()) {

            mirror.add(
                new Position(
                    -1 * position.x,
                    position.y
                )
            );

        }

        this.matrix = mirror;

    }

}
