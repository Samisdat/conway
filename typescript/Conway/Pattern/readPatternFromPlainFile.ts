import {CellMatrix} from '@Conway/Conway/CellMatrix';
import {Position} from '@Conway/Conway/Position';
import {Pattern} from '@Conway/Conway/Pattern';

export const readPatternFromPlainFile = (patternPlainText: string): Pattern => {

    let name: string | undefined = undefined;
    const matrix = new CellMatrix();

    const patternPows: string[] = patternPlainText.split('\n');

    const isComment = new RegExp('^!(.*?)');
    const isMeta = new RegExp('^!(.*?)\:(.*?)$');

    let y = 0;
    let x = 0;

    for (let patternPow of patternPows) {

        patternPow = patternPow.trim();

        if (true === isComment.test(patternPow)) {

            if (true === isMeta.test(patternPow)) {

                const meta = isMeta.exec(patternPow);

                if (null === meta) {
                    continue;
                }

                if ('Name' === meta[1]) {
                    name = meta[2].trim();
                }

            }
            continue;
        }

        const cols: string[] = patternPow.trim().split('');

        x = 0;

        for (const col of cols) {

            if ('O' === col) {

                matrix.add(new Position(x, y));

            }

            x += 1;
        }

        y += 1;

    }

    if (undefined === name) {
        name = 'unnamed';
    }

    const pattern = new Pattern(name, matrix);
    //pattern.mirrorVertical();

    return pattern;

}

