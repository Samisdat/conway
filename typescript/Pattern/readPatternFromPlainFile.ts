import {Position} from '@Conway/Geometry/Position';
import {Pattern} from '@Conway/Pattern/Pattern';
import {Matrix} from '@Conway/Geometry/Matrix';

export const readPatternFromPlainFile = (patternPlainText: string): Pattern => {

    let name: string | undefined = undefined;
    const matrix = new Matrix();

    const patternPows: string[] = patternPlainText.split('\n').reverse();

    const isComment = new RegExp('^!(.*?)');
    const isMeta = new RegExp('^!(.*?)\:(.*?)$');

    let y = 0;
    let x = 0;

    for (let patternPow of patternPows) {

        patternPow = patternPow.trim();

        if (true === isComment.test(patternPow)) {

            if (true === isMeta.test(patternPow)) {

                const meta = isMeta.exec(patternPow) as RegExpMatchArray;

                if ('Name' === meta[1]) {
                    name = meta[2].trim();
                }

            }
            continue;
        }

        const cols: string[] = patternPow.trim().split('');

        x = 0;

        for (const col of cols) {

            if ('O' === col) {

                matrix.add(new Position(x, y));

            }

            x += 1;
        }

        y += 1;

    }

    if (undefined === name) {
        name = 'unnamed';
    }

    const pattern = new Pattern(name, matrix);

    return pattern;

};

