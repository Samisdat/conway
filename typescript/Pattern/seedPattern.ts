import {CellMatrix} from '@Conway/Geometry/CellMatrix';
import {Pattern} from '@Conway/Pattern/Pattern';
import {Position} from '@Conway/Geometry/Position';

export const seedPattern = (matrix: CellMatrix, pattern: Pattern, moveBy = new Position(0, 0)): void => {

    const positions = pattern.getMatrix().all();

    const offSet = new Position(
        -1 * Math.floor(pattern.getMatrix().width() / 2),
        -1 * Math.floor(pattern.getMatrix().height() / 2)
    ).move(moveBy);

    for (const position of positions) {

        matrix.add(position.move(moveBy));

    }

}
