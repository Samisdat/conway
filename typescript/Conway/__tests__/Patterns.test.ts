import {Patterns} from '../Patterns';
import {Position} from '../Position';
import {Pattern} from '../Pattern';
import serializer from '../../../jest-serialize-conway';

describe('Patterns', () => {

    beforeEach(function () {
        expect.addSnapshotSerializer(serializer);
    });

    test('can be created', () => {

        let patterns = new Patterns();

        expect(patterns).toBeInstanceOf(Patterns);

    });

    test('construct fails with unkown pattern', () => {

        let createPattern = () => {

            let patterns = new Patterns();
            patterns.get('foobar');
        };

        expect(createPattern).toThrowErrorMatchingSnapshot();

    });

    const allPatterns = [
        'rotate',
        'block',
        'beehive',
        'loaf',
        'boat',
        'blinker',
        'toad',
        'beacon',
        'pulsar',
        'revolver',
        'glider',
        'lightweight_spaceship',
        'r_pentomino',
        'diehard',
        'acorn',
        'gosper_glider_gun',
        'exhibit_infinite_growth_1',
        'glider_eater',
        'guns_and_eaters',
    ];

    for (const pattern of allPatterns) {

        test('pattern ' + pattern, () => {

            let patterns = new Patterns();

            let blinker = patterns.get(pattern);

            expect(blinker.getMatrix()).toMatchSnapshot();

        });

    }

});
