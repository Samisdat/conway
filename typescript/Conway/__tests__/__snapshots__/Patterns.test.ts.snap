// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Patterns construct fails with unkown pattern 1`] = `"unkown pattern foobar"`;

exports[`Patterns pattern acorn 1`] = `
OXOOOOO
OOOXOOO
XXOOXXX
`;

exports[`Patterns pattern beacon 1`] = `
XXOO
XXOO
OOXX
OOXX
`;

exports[`Patterns pattern beehive 1`] = `
OXXO
XOOX
OXXO
`;

exports[`Patterns pattern blinker 1`] = `
XXX
OOO
`;

exports[`Patterns pattern block 1`] = `
XX
XX
`;

exports[`Patterns pattern boat 1`] = `
XXO
XOX
OXO
`;

exports[`Patterns pattern diehard 1`] = `
OOOOOOXO
XXOOOOOO
OXOOOXXX
`;

exports[`Patterns pattern exhibit_infinite_growth_1 1`] = `
OOOOOOXO
OOOOXOXX
OOOOXOXO
OOOOXOOO
OOXOOOOO
XOXOOOOO
`;

exports[`Patterns pattern glider 1`] = `
OXO
OOX
XXX
`;

exports[`Patterns pattern glider_eater 1`] = `
OOOOOOOXX
OOOXOOOXX
OOXOXOOOO
OXOXOOOOO
OXOOOOOOO
XXOOOOOOO
`;

exports[`Patterns pattern gosper_glider_gun 1`] = `
OOOOOOOOOOOOOOOOOOOOOOOOXOOOOOOOOOOO
OOOOOOOOOOOOOOOOOOOOOOXOXOOOOOOOOOOO
OOOOOOOOOOOOXXOOOOOOXXOOOOOOOOOOOOXX
OOOOOOOOOOOXOOOXOOOOXXOOOOOOOOOOOOXX
XXOOOOOOOOXOOOOOXOOOXXOOOOOOOOOOOOOO
XXOOOOOOOOXOOOXOXXOOOOXOXOOOOOOOOOOO
OOOOOOOOOOXOOOOOXOOOOOOOXOOOOOOOOOOO
OOOOOOOOOOOXOOOXOOOOOOOOOOOOOOOOOOOO
OOOOOOOOOOOOXXOOOOOOOOOOOOOOOOOOOOOO
`;

exports[`Patterns pattern guns_and_eaters 1`] = `
OOOOOOOOOOOXOOOOOOOOOOOOOOOOOOOOOOOOOOOOOXXOOOO
OOOOOOOOOOOXOXOOOOOOOOOOOOOOOOOOOOOOOOOOOXOOOOO
XXOOOOOOOOOOOOXXOOOOOOXXOOOOOOOOOOOOOOOXOXOOOOO
XXOOOOOOOOOOOOXXOOOOXOOOXOOOOOOOOOOOOOXOXOOOOOO
OOOOOOOOOOOOOOXXOOOXOOOOOXOOOOOOOOXXOOOXOOOOOOO
OOOOOOOOOOOXOXOOOOXXOXOOOXOOOOOOOOXXOOOOOOOOOOO
OOOOOOOOOOOXOOOOOOOXOOOOOXOOOOOOOOOOOOOOOOOOOOO
OOOOOOOOOOOOOOOOOOOOXOOOXOOOOOOOOOOOOOOOOOOOOOO
OOOOOOOOOOOOOOOOOOOOOOXXOOOOOOOOOOOOOOOOOOOOOOO
OOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOO
OOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOO
OOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOO
OOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOO
OOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOO
OOOOOOOOOOOOOOOOOOOOOOOXXOOOOOOOOOOOOOOOOOOOOOO
OOOOOOOOOOOOOOOOOOOOOOXOOOXOOOOOOOOOOOOOOOOOOOO
OOOOOOOOOOOOOOOOOOOOOXOOOOOXOOOOOOOXOOOOOOOOOOO
OOOOOOOOOOOXXOOOOOOOOXOOOXOXXOOOOXOXOOOOOOOOOOO
OOOOOOOXOOOXXOOOOOOOOXOOOOOXOOOXXOOOOOOOOOOOOOO
OOOOOOXOXOOOOOOOOOOOOOXOOOXOOOOXXOOOOOOOOOOOOXX
OOOOOXOXOOOOOOOOOOOOOOOXXOOOOOOXXOOOOOOOOOOOOXX
OOOOOXOOOOOOOOOOOOOOOOOOOOOOOOOOOXOXOOOOOOOOOOO
OOOOXXOOOOOOOOOOOOOOOOOOOOOOOOOOOOOXOOOOOOOOOOO
`;

exports[`Patterns pattern lightweight_spaceship 1`] = `
OXOOX
XOOOO
XOOOX
XXXXO
`;

exports[`Patterns pattern loaf 1`] = `
OXXO
XOOX
OXOX
OOXO
`;

exports[`Patterns pattern pulsar 1`] = `
OOXXXOOOXXXOO
OOOOOOOOOOOOO
XOOOOXOXOOOOX
XOOOOXOXOOOOX
XOOOOXOXOOOOX
OOXXXOOOXXXOO
OOOOOOOOOOOOO
OOXXXOOOXXXOO
XOOOOXOXOOOOX
XOOOOXOXOOOOX
XOOOOXOXOOOOX
OOOOOOOOOOOOO
OOXXXOOOXXXOO
`;

exports[`Patterns pattern r_pentomino 1`] = `
OXX
XXO
OXO
`;

exports[`Patterns pattern revolver 1`] = `
XOOOOOOOOOOOOX
XXXOOOOXOOOXXX
OOOXOXOXOOXOOO
OOXOOOOOOXOXOO
OOXOXOOOOOOXOO
OOOXOOXOXOXOOO
XXXOOOXOOOOXXX
XOOOOOOOOOOOOX
`;

exports[`Patterns pattern rotate 1`] = `
XOOO
XOOO
XOOO
XXXX
`;

exports[`Patterns pattern toad 1`] = `
OXXX
XXXO
`;
